specs/capri/thin_opensds_design.md

```diff
@@ -0,0 +1,42 @@
+# Thin OpenSDS Design
+
+**Authors:** [Xing Yang](https://github.com/xing-yang)
+
+This document serves as a design proposal for Thin OpenSDS. This proposal was drafted [here](https://docs.google.com/document/d/16zLCE1X8bNjkGoKvBpqDZaxEHZapWluQVgdvlxuR_tE/edit?usp=sharing).
+
+## Background
+
+In containerized environment, users may not need the full-blown orchestration capabilities of OpenSDS.  Instead, they just want a CSI plugin that can provision storage for Container Orchestration Systems (COs).  That’s why we are proposing to provide a light-weight version of OpenSDS that can serve this need.
+
+## Objectives
+
+### Goals
+
+* Provide a light-weight version of OpenSDS that includes a CSI plugin to provision storage for COs.
+* Only one storage backend is supported at a given time based on configure setting.
+* Only functionalities supported by CSI spec [1] are supported.  These include create/delete volume, attach/detach volume, mount/unmount volume, create/delete snapshot, resize volume, etc.
+* More functionalities will be added to Thin OpenSDS as they become available in the CSI spec.
+
+### Non Goals
+
+* While Thin OpenSDS can support different storage backends, multiple backends cannot be used simultaneously because the scheduler and volume controller are not included.
+* Replication is not supported because CSI does not support replication.
+* Make database configurable and optional. This may be desirable for some use cases, however, it is not easy to achieve as database is needed to persist the states of resources. 
+
+## Design Details
+
+The Thin OpenSDS design is illustrated in the following diagram.
+
+![Thin_OpenSDS Architecture Diagram](thin_opensds.png?raw=true "Thin OpenSDS Architecture Diagram")
+
+This design is dependent upon [API Framework Refactoring](https://github.com/opensds/design-specs/blob/master/specs/capri/API_Framework_Refactoring.md) which has already been implemented in Capri.
```

Comment:
During the implementation for thin-opensds, I found that to make it compatible as much as possible, we should make some decision firstly:
* Currently we use `Metadata` to carry some vendor-specific info so that vendor drivers can provision/attach the volume successfully, so we should make sure `Metadata` contain all required data in apiserver rather than in controller.
* Make status machine a separate package living outside of controller (maybe a struct can be embedded at any time), then apiserver can integrate it more easily when required.

Any thought?

Reply:
Besides, we should make it explicitly for users who want to try thin-opensds that all vendor-specific info (such as poolId, poolName, accessProtocol, etc) should be passed through `Metadata`.

Reply:
This usage of Metadata is not specific to Thin OpenSDS though, isn't it?

Reply:
Yes, in full-opensds it's also used for passing vendor specific data.